c/tskit/tables.c

```diff
@@ -6435,6 +6449,101 @@ tsk_table_collection_check_edge_ordering(tsk_table_collection_t *self)
     return ret;
 }
 
+static int TSK_WARN_UNUSED
+tsk_table_collection_check_mutation_times(tsk_table_collection_t *self)
+{
+    int ret = 0;
+    const tsk_id_t *I, *O;
+    const tsk_edge_table_t edges = self->edges;
+    const tsk_node_table_t nodes = self->nodes;
+    const tsk_site_table_t sites = self->sites;
+    const tsk_mutation_table_t mutations = self->mutations;
+    const tsk_id_t M = (tsk_id_t) edges.num_rows;
+    tsk_id_t tj, tk;
+    tsk_id_t *parent = NULL;
+    double mutation_time, left, right;
+    tsk_id_t site;
+    /* Using unsigned values here avoids potentially undefined behaviour */
+    uint32_t mutation;
+
+    /* We need indexes to step through and find the edges of each mutation to
+       check the parent node time */
+    if (!tsk_table_collection_has_index(self, 0)) {
+        ret = TSK_ERR_TABLES_NOT_INDEXED;
+        goto out;
+    }
+
+    parent = malloc(nodes.num_rows * sizeof(*parent));
+    if (parent == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(parent, 0xff, nodes.num_rows * sizeof(*parent));
+
+    I = self->indexes.edge_insertion_order;
+    O = self->indexes.edge_removal_order;
+    tj = 0;
+    tk = 0;
+    site = 0;
+    mutation = 0;
+    left = 0;
+    while (tj < M || left < self->sequence_length) {
+        while (tk < M && edges.right[O[tk]] == left) {
+            parent[edges.child[O[tk]]] = TSK_NULL;
+            tk++;
+        }
+        while (tj < M && edges.left[I[tj]] == left) {
+            parent[edges.child[I[tj]]] = edges.parent[I[tj]];
+            tj++;
+        }
+        right = self->sequence_length;
+        if (tj < M) {
+            right = TSK_MIN(right, edges.left[I[tj]]);
+        }
+        if (tk < M) {
+            right = TSK_MIN(right, edges.right[O[tk]]);
+        }
+
+        /* Tree is now ready. We look at each site on this tree in turn */
+        while (site < (tsk_id_t) sites.num_rows && sites.position[site] < right) {
+            while (mutation < mutations.num_rows && mutations.site[mutation] == site) {
+                mutation_time = mutations.time[mutation];
+                if (mutation_time == TSK_NULL) {
+                    ret = TSK_ERR_NULL_TIME;
```

Comment:
Gee, I at least think of TSK_NULL as the integer label -1, not a float. Should this instead check for negative times, and if not, say that they must be nonnegative?

Reply:
I think checking for non-negative is the right thing to do, so will switch it to that.

Reply:
Ok didn't realise node times could be negative, I will remove `TSK_ERR_NULL_TIME` as an error completely, and -1 will just be a default where needed.

Reply:
`TSK_ERR_NULL_TIME` removed in https://github.com/tskit-dev/tskit/pull/672/commits/06e1d3787436bec2823f8d92b18cfbd79715c0bc

Reply:
We do? Oh, that is vaguely familiar. Happliy, our [documentation agrees](https://tskit.readthedocs.io/en/latest/data-model.html#node-requirements).